tests/junit.py

```diff
@@ -169,17 +178,40 @@ def run(test, backends):
 
         result = run(args.test, backends)
 
-        junit_batch = ''
-        try:
-            junit_batch = '-' + os.environ['JUNIT_BATCH']
-        except:
-            pass
-        output = (os.path.join('build', args.test + junit_batch + '.junit')
-                  if args.output is None
-                  else args.output)
-
-        with open(output, 'w') as fd:
-            TestSuite.to_file(fd, [result])
+        if args.mode.lower() == "junit":
+            junit_batch = ''
+            try:
+                junit_batch = '-' + os.environ['JUNIT_BATCH']
+            except:
+                pass
+            output = (os.path.join('build', args.test + junit_batch + '.junit')
+                      if args.output is None
+                      else args.output)
+
+            with open(output, 'w') as fd:
+                TestSuite.to_file(fd, [result])
+        elif args.mode.lower() == "tap":
+            print('1..' + str(len(result.test_cases)))
+            for i in range(len(result.test_cases)):
+                test_case = result.test_cases[i]
+                test_index = str(i + 1)
+                print('# Test: ' + test_case.name.split(' ')[1])
+                print('# $ ' + test_case.args)
+                if test_case.is_error():
+                    message = test_case.is_error() if isinstance(test_case.is_error(), str) else test_case.errors[0]['message']
+                    print('not ok {} - ERROR: {}'.format(test_index, message.strip()))
+                    print(test_case.errors[0]['output'].strip())
+                elif test_case.is_failure():
+                    message = test_case.is_failure() if isinstance(test_case.is_failure(), str) else test_case.failures[0]['message']
+                    print('not ok {} - FAIL: {}'.format(test_index, message.strip()))
+                    print(test_case.failures[0]['output'].strip())
+                elif test_case.is_skipped():
+                    message = test_case.is_skipped() if isinstance(test_case.is_skipped(), str) else test_case.skipped[0]['message']
+                    print('ok {} - SKIP: {}'.format(test_index, message.strip()))
```

Comment:
I had to update my logic on handling these cases in Ratel due to a misunderstanding on my part: https://gitlab.com/micromorph/ratel/-/merge_requests/235
Note that the change included updating to a newer version of the Python package we're using here

Reply:
Should I add all the changes in that MR to this branch?

Reply:
I'd recommend that. You can copy-paste the tests/junit-xml folder over from Ratel main and then that let's you use those changes from junit.py

Reply:
I copied from ratel. In this branch I get
```
make prove -j PROVE_OPTS=-v

All tests successful.
Files=208, Tests=1744, 28 wallclock secs ( 1.23 usr  0.36 sys + 19.61 cusr  5.44 csys = 26.64 CPU)
Result: PASS
```

But in main
we have more tests, why?
```
All tests successful.
Files=208, Tests=5232, 53 wallclock secs ( 3.27 usr  0.69 sys + 27.00 cusr 26.10 csys = 57.06 CPU)
Result: PASS
```

Reply:
5283/3 = 1744

Currently we report each test in triplicate, stdout, stderr, and the return code, but in Ratel we combined to a single report per test case

Reply:
Nice!